README.txt
Introduction
============

This Script is mainly a shell command generator which executes the command in
the parent shell.

Requirements
------------

This script needs Python 2.7 because of argparse and depends on appscript.

Installation
------------

The best thing is to install it in a Python 2.7 virtualenv for testing::

    $ virtualenv -p python2.7 --no-site-packages pcmd
    $ . pcmd/bin/activate
    $ pip install --editable=git+https://github.com/ramonski/pcmd#egg=pcmd

Configuration
-------------

You can configure your **p-commands** in **$HOME/.p.ini**.

Syntax::

    [<p-section>]
    <p-command> = <any arbitrary shell command>

Call::

    $ p <p-command> <p-section>

Example::

    [develop]
    goto = cd /Users/rbartl/develop
    mvim = mvim /Users/rbartl/develop

The p-command executes the shell command by default in the **Terminal.app**.
This can be overwritten with the Environment Variable **PTERM**, e.g.::

    $ vim .bashrc
    export PTERM="iTerm"


Usage
-----

After you configured some commands, you can execute them like this::

    $ p goto develop # executes -> cd /Users/rbartl/develop
    $ p mvim develop # executes -> mvim /Users/rbartl/develop

Bash Completion
---------------

The p-command supports now bash completion.
If you use bash, you can source the pcmd_wrapper.sh to enable bash completion::

    $ source pcmd_wrapper.sh
    $ p o<TAB>
    option1 option2

    $ p option1 s<TAB>
    section1 section2


docs/HISTORY.txt
Changelog
=========

0.2 - 2011-06-05
----------------

  - factored out the ini config part
    [ramonski]

  - added bash completion script
    [ramonski]


0.1 - 2011-06-03
----------------

  - initial release
    [ramonski]


pcmd/config.py
# -*- coding: utf-8 -*-
#
# File: config.py
#
# Copyright (c) ramonski
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = 'Ramon Bartl <ramon_bartl@yahoo.de>'
__docformat__ = 'plaintext'

import os
import ConfigParser

INIFILE = ".p.ini"

def get_config():
    home = os.environ.get("HOME")
    return os.sep.join([home, INIFILE])

def is_config_there():
    return os.path.exists(get_config())

def get_parser():
    parser = ConfigParser.SafeConfigParser()
    parser.read(get_config())
    return parser

def sections():
    parser = get_parser()
    return parser.sections()

def options():
    parser = get_parser()
    return [parser.options(s) for s in parser.sections()]

def _sections():
    for section in sorted(sections()):
        print section

def _options():
    out = []
    _ = [out.extend(o) for o in options()]
    for option in out:
        print option

# vim: set ft=python ts=4 sw=4 expandtab :


pcmd/__init__.py
# -*- coding: utf-8 -*-
#
# File: __init__.py
#
# Copyright (c) ramonski
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = 'Ramon Bartl <ramon_bartl@yahoo.de>'
__docformat__ = 'plaintext'

import os
import argparse
import appscript

from config import get_parser
from config import get_config
from config import is_config_there

TERMINAL_ENV = "PTERM"
TERMINAL_APP = "Terminal"


def terminal():
    terminal = os.environ.get(TERMINAL_ENV, TERMINAL_APP)
    app = appscript.app(terminal)
    return app


def active_terminal_window():
    app = terminal()
    for window in app.windows():
        if not window.frontmost():
            continue
        return window


def active_terminal_tab():
    win = active_terminal_window()
    return win.selected_tab()


def do_script(cmd):
    tab = active_terminal_tab()
    return tab.do_script(cmd, in_=tab)


def main():
    parser = argparse.ArgumentParser(description='Runs a shell command in the parent shell')

    parser.add_argument('cmd',
                        metavar='COMMAND',
                        type=str,
                        help='Command to execute')

    parser.add_argument('section',
                        metavar='SECTION',
                        type=str,
                        help='Section where the Command is executed')

    args = parser.parse_args()

    pcmd = get_parser()

    if not is_config_there():
        msg = "No Config found in %s\n" % get_config()
        parser.exit(status=1, message=msg)

    if not pcmd.has_section(args.section):
        msg = "Section %s not in config\n" % args.section
        parser.exit(status=1, message=msg)

    if not pcmd.has_option(args.section, args.cmd):
        msg = "Command %s not in Section %s\n" % (args.cmd,  args.section)
        parser.exit(status=1, message=msg)

    # run the shell command in the parent terminal
    do_script(pcmd.get(args.section, args.cmd))


if __name__ == '__main__':
    main()

# vim: set ft=python ts=4 sw=4 expandtab :
